File: LeetCode-solution-Java/account_Merge.java

import java.util.*;
import java.io.*;
import java.util.Map.Entry;

class Solution {
    int[] parent;
    int[] rank;

    private int findParent(int u) {
        if (u == parent[u])
            return u;
        return parent[u] = findParent(parent[u]);
    }

    private void union(int u, int v) {
        int U = findParent(u);
        int V = findParent(v);
        if (rank[U] < rank[V]) {
            parent[U] = V;
        } else if (rank[V] < rank[U]) {
            parent[V] = U;
        } else {
            parent[V] = U;
            rank[U]++;
        }
    }

    public List<List<String>> accountsMerge(List<List<String>> accounts) {
        HashMap<String, Integer> emailsToInts = new HashMap<>();
        HashMap<Integer, String> intsToEmails = new HashMap<>();
        int x = 0;
        int n = accounts.size();
        for (int i = 0; i < n; i++) {
            List<String> account = accounts.get(i);
            int m = account.size();
            for (int j = 1; j < m; j++) {
                String email = account.get(j);
                boolean alreadyMapped = emailsToInts.containsKey(email);
                if (alreadyMapped)
                    continue;
                emailsToInts.put(email, x);
                intsToEmails.put(x++, email);
            }
        }
        String[] names = new String[x];
        ArrayList<ArrayList<Integer>> connections = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            List<String> account = accounts.get(i);
            String name = account.get(0);
            int m = account.size();
            ArrayList<Integer> temp = new ArrayList<>();
            for (int j = 1; j < m; j++) {
                int ind = emailsToInts.get(account.get(j));
                names[ind] = name;
                temp.add(ind);
            }
            connections.add(temp);
        }
        parent = new int[x];
        rank = new int[x];
        for (int i = 0; i < x; i++)
            parent[i] = i;

        for (ArrayList<Integer> connection : connections) {
            int z = connection.size();
            for (int i = 0; i < z; i++) {
                for (int j = i + 1; j < z; j++) {
                    union(connection.get(i), connection.get(j));
                }
            }
        }

        for (int i = 0; i < x; i++)
            parent[i] = findParent(i);

        HashMap<Integer, TreeSet<String>> map = new HashMap<>();
        for (int i = 0; i < x; i++) {
            if (!map.containsKey(parent[i])) {
                map.put(parent[i], new TreeSet<String>());
            }
            map.get(parent[i]).add(intsToEmails.get(i));
        }
        List<List<String>> ans = new ArrayList<>();
        for (Entry<Integer, TreeSet<String>> entry : map.entrySet()) {
            String name = names[entry.getKey()];
            TreeSet<String> emails = entry.getValue();
            List<String> temp = new ArrayList<>();
            temp.add(name);
            for (String email : emails) {
                temp.add(email);
            }
            ans.add(temp);
        }
        return ans;
    }
}

File: LeetCode-Solutions-Kotlin/accountMerge.kt
    import java.util.*
    import kotlin.collections.HashMap

    fun findParent(u: String, parents: HashMap<String, String>): String {
        if (parents[u].equals(u)) return u
        parents[u] = findParent(parents[u]!!, parents)
        return parents[u]!!
    }

    fun union(u: String, v: String, parents: HashMap<String, String>, ranks: HashMap<String, Int>) {
        val parU = findParent(u, parents)
        val parV = findParent(v, parents)
        if (parU == parV) {
            return
        } else if (ranks[parU]!! < ranks[parV]!!) {
            parents[parU] = parV
        } else if (ranks[parV]!! < ranks[parU]!!) {
            parents[parV] = parU
        } else {
            parents[parV] = parU
            ranks[parU]!!.inc()
        }
    }

    fun _accountsMerge(accounts: List<List<String>>): List<List<String>> {
        val parents = HashMap<String, String>()
        val ranks = HashMap<String, Int>()
        val emailToNames = HashMap<String, String>()
        val namesToEmails = HashMap<String, TreeSet<String>>()

        for (account in accounts) {
            val n = account.size
            for (i in 1 until n) {
                parents[account[i]] = account[i]
                ranks[account[i]] = 0
                emailToNames[account[i]] = account[0]
            }
        }

        for (account in accounts) {
            val n = account.size
            for (i in 1 until n) {
                for (j in i + 1 until n) {
                    union(account[i], account[j], parents, ranks)
                }
            }
        }

        parents.keys.forEach {
            val name = findParent(it, parents)
            if(!namesToEmails.containsKey(name)) namesToEmails[name] = TreeSet()
            namesToEmails[name]!!.add(it)
        }


        val ans = ArrayList<ArrayList<String>>()
        for (record in namesToEmails) {
            val data = ArrayList<String>()
            data.add(emailToNames[record.key]!!)
            record.value.forEach { data.add(it) }
            ans.add(data)
        }

        return ans
    }

    fun accountsMerge(accounts: List<List<String>>): List<List<String>> {
        var group = 0
        val groups = HashMap<String, Int>()
        val names = HashMap<Int, String>()
        accounts.forEach { account ->
            val n = account.size
            names[group] = account[0]
            var oldGroup = -1
            for(i in 1 until n) {
                if(groups.containsKey(account[i])){
                    oldGroup = groups[account[i]]!!
                    break
                }else{
                    groups[account[i]] = group
                    group++
                }
            }
            if(oldGroup != -1) {
                for(i in 1 until n) {
                    groups[account[i]] = oldGroup
                }
            }
        }

        val groupEmails = HashMap<Int, TreeSet<String>>()
        groups.forEach { (email, group) ->
            if(!groupEmails.containsKey(group)) groupEmails[group] = TreeSet()
            groupEmails[group]!!.add(email)
        }

        val ans = ArrayList<ArrayList<String>>()
        groupEmails.forEach { (grp, emails) ->
            val temp = ArrayList<String>()
            temp.add(names[grp]!!)
            emails.forEach { temp.add(it) }
            ans.add(temp)
        }
        return ans
    }